File: backend/app/games/mendikot/ai.py
"""Mendikot bot — port of the frontend heuristic AI."""
from __future__ import annotations

from typing import List, Optional

from ..kachuful.deck import RANK_VALUE
from ..kachuful.types import Card, Suit
from .engine import (
    active_trump,
    legal_cards,
    player_at_seat,
    suit_counts,
    team_of,
    trick_winner_seat,
)
from .types import MendikotState

SUITS: List[Suit] = ["S", "H", "D", "C"]


def _is_ten(c: Card) -> bool:
    return c.rank == "10"


def _lowest(cards: List[Card]) -> Card:
    return sorted(cards, key=lambda c: RANK_VALUE[c.rank])[0]


def _highest(cards: List[Card]) -> Card:
    return sorted(cards, key=lambda c: -RANK_VALUE[c.rank])[0]


def bot_choose_trump(hand: List[Card]) -> Suit:
    counts = suit_counts(hand)
    best: Suit = "S"
    best_score = -1
    for s in SUITS:
        strength = sum(RANK_VALUE[c.rank] for c in hand if c.suit == s)
        score = counts[s] * 100 + strength
        if score > best_score:
            best_score = score
            best = s
    return best


def bot_play(state: MendikotState) -> Card:
    seat = state.turnSeat
    my_team = team_of(seat)
    legal = legal_cards(state, seat)
    trump = active_trump(state)

    # Leading.
    if not state.currentTrick:
        non_tens = [c for c in legal if not _is_ten(c)]
        pool = non_tens if non_tens else legal
        off_trump = [c for c in pool if c.suit != state.trump]
        candidates = off_trump if off_trump else pool
        aces = [c for c in candidates if c.rank == "A"]
        if aces:
            return aces[0]
        return _highest(candidates)

    lead_suit = state.leadSuit
    winning_seat = trick_winner_seat(state.currentTrick, trump, lead_suit)
    partner_winning = team_of(winning_seat) == my_team
    trick_has_ten = any(_is_ten(t.card) for t in state.currentTrick)
    win_card = next(t.card for t in state.currentTrick if t.seat == winning_seat)
    can_follow = any(c.suit == lead_suit for c in legal)

    if can_follow:
        suit_cards = [c for c in legal if c.suit == lead_suit]
        if partner_winning:
            my_ten = next((c for c in suit_cards if _is_ten(c)), None)
            if my_ten:
                return my_ten
            return _lowest(suit_cards)
        beats = [
            c for c in suit_cards if RANK_VALUE[c.rank] > RANK_VALUE[win_card.rank]
        ]
        if beats:
            non_ten_beats = [c for c in beats if not _is_ten(c)]
            if trick_has_ten or not non_ten_beats:
                return _lowest(beats)
            return _lowest(non_ten_beats)
        non_tens = [c for c in suit_cards if not _is_ten(c)]
        return _lowest(non_tens if non_tens else suit_cards)

    # Void in lead suit.
    trumps_in_hand = [c for c in legal if trump and c.suit == trump]
    if (not partner_winning) and trumps_in_hand:
        if win_card.suit != trump:
            return _lowest(trumps_in_hand)
        beating = [
            c for c in trumps_in_hand if RANK_VALUE[c.rank] > RANK_VALUE[win_card.rank]
        ]
        if beating:
            return _lowest(beating)
    non_tens = [c for c in legal if not _is_ten(c)]
    return _lowest(non_tens if non_tens else legal)


def bot_action(state: MendikotState, player_id: str) -> Optional[dict]:
    if state.phase == "choose_trump":
        hand = state.hands.get(player_id, [])
        return {"type": "chooseTrump", "suit": bot_choose_trump(hand)}
    if state.phase == "playing":
        card = bot_play(state)
        return {"type": "play", "cardId": card.id}
    return None


def auto_default_action(state: MendikotState, player_id: str) -> Optional[dict]:
    """Fallback action when a human's turn times out."""
    if state.phase == "choose_trump":
        hand = state.hands.get(player_id, [])
        return {"type": "chooseTrump", "suit": bot_choose_trump(hand)}
    if state.phase == "playing":
        seat = player_at_seat(state, state.turnSeat).seat
        legal = legal_cards(state, seat)
        if not legal:
            return None
        return {"type": "play", "cardId": _lowest(legal).id}
    return None

File: src/games/mendikot/engine/ai.ts
import { RANK_VALUE } from "../../kachuful/engine/deck";
import {
  activeTrump,
  legalCards,
  suitCounts,
  trickWinnerSeat,
} from "./engine";
import { teamOf, type Card, type MendikotState, type Suit } from "./types";

const SUITS: Suit[] = ["S", "H", "D", "C"];

function isTen(c: Card) {
  return c.rank === "10";
}
function lowest(cards: Card[]) {
  return [...cards].sort((a, b) => RANK_VALUE[a.rank] - RANK_VALUE[b.rank])[0];
}
function highest(cards: Card[]) {
  return [...cards].sort((a, b) => RANK_VALUE[b.rank] - RANK_VALUE[a.rank])[0];
}

/** Bot picks trump = its longest suit (tie-break by total rank strength). */
export function botChooseTrump(hand: Card[]): Suit {
  const counts = suitCounts(hand);
  let best: Suit = "S";
  let bestScore = -1;
  for (const s of SUITS) {
    const strength = hand
      .filter((c) => c.suit === s)
      .reduce((sum, c) => sum + RANK_VALUE[c.rank], 0);
    const score = counts[s] * 100 + strength;
    if (score > bestScore) {
      bestScore = score;
      best = s;
    }
  }
  return best;
}

/** Choose a card for the bot whose turn it is. */
export function botPlay(s: MendikotState): Card {
  const seat = s.turnSeat;
  const myTeam = teamOf(seat);
  const legal = legalCards(s, seat);
  const trump = activeTrump(s);

  // Leading the trick.
  if (s.currentTrick.length === 0) {
    const nonTens = legal.filter((c) => !isTen(c));
    const pool = nonTens.length > 0 ? nonTens : legal;
    // Lead a strong card (Ace/King) to pull tens; prefer non-trump leads.
    const offTrump = pool.filter((c) => c.suit !== s.trump);
    const candidates = offTrump.length > 0 ? offTrump : pool;
    const aces = candidates.filter((c) => c.rank === "A");
    if (aces.length > 0) return aces[0];
    return highest(candidates);
  }

  // Following.
  const leadSuit = s.leadSuit!;
  const winningSeat = trickWinnerSeat(s.currentTrick, trump, leadSuit);
  const partnerWinning = teamOf(winningSeat) === myTeam;
  const trickHasTen = s.currentTrick.some((t) => isTen(t.card));
  const canFollow = legal.some((c) => c.suit === leadSuit);

  if (canFollow) {
    const suitCards = legal.filter((c) => c.suit === leadSuit);
    const winCard = s.currentTrick.find((t) => t.seat === winningSeat)!.card;

    if (partnerWinning) {
      // Partner safe → dump a ten of this suit to bank it, else play low.
      const myTen = suitCards.find((c) => isTen(c));
      if (myTen) return myTen;
      return lowest(suitCards);
    }
    // Opponent winning → try to beat (cheaply); secure tens for our team.
    const beats = suitCards.filter(
      (c) => RANK_VALUE[c.rank] > RANK_VALUE[winCard.rank]
    );
    if (beats.length > 0) {
      // win with lowest sufficient card, preferring not to waste our ten
      const nonTenBeats = beats.filter((c) => !isTen(c));
      if (trickHasTen || nonTenBeats.length === 0) return lowest(beats);
      return lowest(nonTenBeats);
    }
    // can't beat → throw lowest non-ten
    const nonTens = suitCards.filter((c) => !isTen(c));
    return lowest(nonTens.length > 0 ? nonTens : suitCards);
  }

  // Void in lead suit.
  const trumpsInHand = trump ? legal.filter((c) => c.suit === trump) : [];
  if (!partnerWinning && trumpsInHand.length > 0) {
    // ruff to win (especially if a ten is at stake)
    const trumpsBeating = trumpsInHand.filter((c) => {
      const winCard = s.currentTrick.find((t) => t.seat === winningSeat)!.card;
      if (winCard.suit !== trump) return true; // any trump beats a non-trump
      return RANK_VALUE[c.rank] > RANK_VALUE[winCard.rank];
    });
    if (trumpsBeating.length > 0) return lowest(trumpsBeating);
  }
  // discard lowest non-ten (keep tens out of opponents' hands)
  const nonTens = legal.filter((c) => !isTen(c));
  return lowest(nonTens.length > 0 ? nonTens : legal);
}
